### PR TITLE
Enable in-tree build

### PR DESCRIPTION
**Pull request by mgehre-amd**
This allows to build onnx-mlir in-tree with LLVM, e.g. by doing

```
docker run --rm --user $(id -u):$(id -g) -i -v $PWD:/workdir/onnx-mlir -w /workdir/onnx-mlir onnxmlir/onnx-mlir-dev << EOF
bash utils/clone-mlir.sh
rm -rf build
cmake llvm-project/llvm \
  -Bbuild \
  -G Ninja \
  -DCMAKE_BUILD_TYPE=Release \
  -DLLVM_ENABLE_PROJECTS=mlir \
  -DLLVM_EXTERNAL_PROJECTS="onnx-mlir" \
  -DONNX_MLIR_ENABLE_STABLEHLO=OFF \
  -DLLVM_ENABLE_RTTI=ON \
  -DLLVM_EXTERNAL_ONNX_MLIR_SOURCE_DIR=. \
  -DLLVM_TARGETS_TO_BUILD=host \
  -DLLVM_BUILD_TOOLS=OFF \
  -DLLVM_BUILD_UTILS=OFF \
  -DLLVM_BUILD_RUNTIMES=OFF

ninja -C build onnx-mlir onnx-mlir-opt
EOF
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,14 +55,16 @@ if (ONNX_MLIR_CCACHE_BUILD)
   endif()
 endif()
 
-# Enable warnings as errors
-# Leverage the imported LLVM_ENABLE_WERROR for compiler logic
-set(LLVM_ENABLE_WERROR ${ONNX_MLIR_ENABLE_WERROR})
+if (CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
```
Comment (TinaAMD) [inline CMakeLists.txt:58]: Maybe we can use a function to identify in-tree builds? I think this (or the negated version) is used various times.

File: CMakeLists.txt [hunk 2]
```diff
@@ -166,10 +172,14 @@ add_subdirectory(third_party/pybind11)
 add_subdirectory(third_party/rapidcheck)
 
 if (ONNX_MLIR_ENABLE_STABLEHLO)
+  if (ONNX_MLIR_BUILD_INTREE)
+    message(FATAL_ERROR "In tree builds don't support stablehloe yet. "
```
Comment (cferry-AMD) [inline CMakeLists.txt]:
```suggestion
    message(FATAL_ERROR "In tree builds don't support stablehlo yet. "
```

File: CMakeLists.txt (hunk 1 above)
Comment (mgehre-amd) [reply CMakeLists.txt:58]: done

File: .github/workflows/ubuntu-build-intree.yml [hunk 3]
```diff
@@ -0,0 +1,102 @@
+name: Build in-tree
```
Comment (TinaAMD) [inline .github/workflows/ubuntu-build-intree.yml]:
```suggestion
name: In-tree build
```
(to align with the other action)

File: .github/workflows/ubuntu-build-intree.yml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+on:
+  pull_request:
+  push:
+    branches: [ main, feature/onnx-to-tosa ]
+
+concurrency:
+  # Build every push to main
```
Comment (TinaAMD) [inline .github/workflows/ubuntu-build-intree.yml]:
```suggestion
```

File: .github/workflows/ubuntu-build-intree.yml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        python-version: '3.10'
+
+    - name: install tools that are needed for compilation
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y gcc g++ cmake ninja-build
+    - name: Setup ccache
+      uses: hendrikmuhs/ccache-action@v1
+      with:
+        # A full build of llvm, clang, lld, and lldb takes about 250MB
```
Comment (TinaAMD) [inline .github/workflows/ubuntu-build-intree.yml]: We do not enable `clang`/`lld`/`lldb` below in `LLVM_ENABLE_PROJECTS`, right? Is this comment copied from somewhere?

File: .github/workflows/ubuntu-build-intree.yml [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        cd protobuf
+        ./autogen.sh
+        ./configure --enable-static=no --prefix=/usr
+        sudo make -j2 install
+
+    - name: install python requirements
+      run: |
+        python3 -m pip install --upgrade wheel
+        python3 -m pip install -r requirements.txt
+        pip install onnx==1.15.0
```
Comment (TinaAMD) [inline .github/workflows/ubuntu-build-intree.yml]: Should we have the onnx version in a file somewhere else? Maybe also the protobuf version?

File: test/mlir/driver/version.mlir [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // RUN: onnx-mlir --version | FileCheck %s
 
 // CHECK:   onnx-mlir version {{[0-9]+}}.{{[0-9]+}}.{{[0-9]+}}, onnx version {{[0-9]+}}.{{[0-9]+}}.{{[0-9]+}}
-// CHECK:   LLVM version {{[0-9]+}}.{{[0-9]+}}.{{[0-9]+}}
+// CHECK:   LLVM version {{.*}}
```
Comment (TinaAMD) [inline test/mlir/driver/version.mlir]: What will this look like after the PR? Will it be empty depending on the in-tree/oot build?

File: .github/workflows/ubuntu-build-intree.yml [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      run: |
+        export LIT_OPTS=-v
+        cmake --build build --target check-onnx-lit
+
+
+    - name: build and run docs/doc_example tests
+      run: |
+        cd ..
+        sh onnx-mlir/utils/check-doc-example.sh
+
```
Comment (TinaAMD) [inline .github/workflows/ubuntu-build-intree.yml:82]: Maybe we can share some steps between both actions?

File: src/Compiler/CMakeLists.txt [hunk 9]
```diff
@@ -69,6 +71,7 @@ add_onnx_mlir_library(OMCompilerDialects
   OMONNXOps
   MLIRIR
   MLIROpenMPToLLVMIRTranslation
+  ${dialect_libs}
```
Comment (cferry-AMD) [inline src/Compiler/CMakeLists.txt:74]: I'm a bit unsure about why we need to add this now: how were these linked to ONNX-MLIR before?

File: .github/workflows/ubuntu-build-intree.yml [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    - name: build
+      run: |
+        cmake llvm-project/llvm \
+          -Bbuild \
+          -G Ninja \
+          -DCMAKE_BUILD_TYPE=Release \
+          -DLLVM_ENABLE_PROJECTS=mlir \
+          -DLLVM_EXTERNAL_PROJECTS="onnx-mlir" \
+          -DONNX_MLIR_ENABLE_STABLEHLO=OFF \
+          -DLLVM_EXTERNAL_ONNX_MLIR_SOURCE_DIR=. \
```
Comment (cferry-AMD) [inline .github/workflows/ubuntu-build-intree.yml:56]: In case of an in-tree build, shouldn't this be inferred?

**Review by cferry-AMD — APPROVED** [review]
It's a good idea, I do have a few questions though.

File: src/Compiler/CMakeLists.txt (hunk 9 above)
Comment (mgehre-amd) [reply src/Compiler/CMakeLists.txt:74]: in-tree builds  require the dependencies to be setup correctly. Before, we would always build all of LLVM before onnx-mlir, so every library had all generated code (from tablegen) available anyways, without declaring any dependencies.
Now we need to declare them correctly to ensure that the tablegen runs before the generated headers are used.

File: .github/workflows/ubuntu-build-intree.yml (hunk 10 above)
Comment (mgehre-amd) [reply .github/workflows/ubuntu-build-intree.yml:56]: In-tree just means that both llvm and onnx-mlir are within the same ninja build graph. It doesn't mean that llvm will magically know where the other projects are.

File: .github/workflows/ubuntu-build-intree.yml (hunk 5 above)
Comment (mgehre-amd) [reply .github/workflows/ubuntu-build-intree.yml]: Yes, I think it's from the llvm-project repo. I will adjust the comment

File: src/Compiler/CMakeLists.txt (hunk 9 above)
Comment (cferry-AMD) [reply src/Compiler/CMakeLists.txt:74]: Right, thanks for the explanation!

File: .github/workflows/ubuntu-build-intree.yml (hunk 6 above)
Comment (mgehre-amd) [reply .github/workflows/ubuntu-build-intree.yml]: Ok, I'll move it.

File: test/mlir/driver/version.mlir (hunk 7 above)
Comment (mgehre-amd) [reply test/mlir/driver/version.mlir]: Turns out I didn't need this with a correct fix: 54794c38e

File: .github/workflows/ubuntu-build-intree.yml (hunk 8 above)
Comment (mgehre-amd) [reply .github/workflows/ubuntu-build-intree.yml:82]: Shared in 90c226722

**Review by TinaAMD — APPROVED** [review]
LGTM

**Pull request merged**